asdf/_block/external.py

```diff
@@ -0,0 +1,42 @@
+"""
+For external blocks, the previous block management
+would cache data opened from external files (to return the
+same underlying ndarray if the same external block
+was referenced more than once). `ExternalBlockCache` is
+used here to allow for the same behavior without requiring
+the block manager to have a reference to the `AsdfFile`
+(that references the block manager).
+"""
+import os
+
+from asdf import generic_io, util
+
+
+class UseInternal:
+    pass
+
+
+class ExternalBlockCache:
+    def __init__(self):
+        self._cache = {}
```

Comment:
I'm not quite sure about this one.

The old code called `AsdfFile.open_external` which caches every opened AsdfFile (in a dictionary). This means that a repeat call to `open_external` (and then accessing the first block) always returns the same array as `Block` caches the data). This behavior is relied upon for matching arrays that share an external block as tested in `test_explode_then_implode`:
https://github.com/asdf-format/asdf/blob/b0c9a50e8d3add337e1271369a3bf834925176a4/asdf/_tests/commands/tests/test_exploded.py#L11-L51
Changing `ExternalBlockCache` to LRU would mean that when the cache overflows, a call to `load` would return an array with a different id which could interfere with settings assignment and block sharing.

I'm inclined to view the memory mapping of external blocks as a bug (and fix it in this PR by not memory mapping) but that does create an issue for large exploded files as you pointed out. As `ExternalBlockCache` is used by the `Manger` that is aware of if the blocks should be lazy loaded and/or memory mapped another option would be to see how complicated it would be to pass these options to `ExternalBlockCache.load` and allow the user some control over how the external blocks are loaded. What do you think about this as an option for addressing the concern about large exploded files?

Reply:
I spent a bit of time looking at this today and I think there may be a few issues with passing along open options. 

`lazy_load` is not an issue (and is [currently supported in this PR](https://github.com/braingram/asdf/blob/642fd7f1c244d9d7662a5e5563e2a98f6f30e358/asdf/core/_converters/ndarray.py#L162-L181) as the ndarray converter sets up for lazy loading the external block then forces it's loading if `lazy_load` was disabled).

`validate_checksums` also should not be an issue but is not currently being passed on for this PR.

`copy_arrays` is where things get complicated. The main complication is when the external file is closed. For a non-memmapped external block, the external file can be opened, the block data read and the file closed before the block data is returned. This is not true for memmapped blocks where the external file will need to be held open until the `AsdfFile` instance is closed. This was one of the reasons why the previous `BlockManager` held a reference to the `AsdfFile` (so that it could used `AsdfFile.open_external` which can keep the external files open and close them when the `AsdfFile` is closed).

I could see a few options to make memory mapping external blocks possible but I'm not sure if the complexity outweighs the benefits.

1. Duplicate the `AsdfFile.open_external` functionality in `_block.Manager`. asdf prior to this PR called `close` on the block manager and we will need to re-add that call to allow the block manager to close the external files.
2. Allow for generating pure `numpy.memmap` objects (which don't guarantee file closing) for external blocks.

I'm going to think about this one some more but any suggestions on how to handle this are appreciated.